File: scala/Parsing/AST.scala
package Parsing

/**
 * romanesco抽象構文木（AST）
 */

/** 式 */
enum Expr:
  case Num(value: String)                           // 数値リテラル
  case Var(name: String)                            // 変数参照
  case Call(f: Expr, args: List[Expr])              // 関数呼び出し
  case Block(exprs: List[Expr])                     // ブロック式
  case Lambda(param: String, body: Expr)            // ラムダ式

/** ステートメント */
enum Stmt:
  case ExprStmt(expr: Expr)                                     // 式文

// ======================================
// Pretty Printer
// ======================================

object Expr:
  /** 式を文字列表現に変換 */
  def show(expr: Expr): String = expr match
    case Num(v) => v
    case Var(n) => n
    case Call(f, args) => s"${show(f)}(${args.map(show).mkString(", ")})"
    case Block(exprs) => s"{ ${exprs.map(show).mkString("; ")} }"
    case Lambda(param, body) => s"\\$param -> ${show(body)}"

object Stmt:
  /** ステートメントを文字列表現に変換 */
  def show(stmt: Stmt): String = stmt match
    case ExprStmt(e) => Expr.show(e)

File: Python/romanesco/syntax/ast.py
from dataclasses import dataclass
from typing import List, Any

# ======================================
# AST Nodes
# ======================================

class Expr: pass

@dataclass
class Num(Expr):
    value: str
    def __repr__(self): return f"Num({self.value})"

@dataclass
class Var(Expr):
    name: str
    def __repr__(self): return f"Var({self.name})"

@dataclass
class Call(Expr):
    f: Expr
    args: List[Expr]
    def __repr__(self): return f"Call({self.f}, {self.args})"

@dataclass
class Block(Expr):
    exprs: List[Expr]
    def __repr__(self): return f"Block({self.exprs})"

@dataclass
class Lambda(Expr):
    param: str
    body: Expr
    def __repr__(self): return f"Lambda({self.param}, {self.body})"

# For internal engine use or simple core mapping
@dataclass
class Atom(Expr):
    name: str
    def __repr__(self): return f"Atom({self.name})"

@dataclass
class Apply(Expr):
    f: Any
    arg: Any
    def __repr__(self): return f"Apply({self.f}, {self.arg})"

class Stmt: pass

@dataclass
class ExprStmt(Stmt):
    expr: Expr
